File: samples/extism-vert-x/src/main/java/garden/bots/starter/MainVerticle.java
package garden.bots.starter;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Promise;

import org.extism.sdk.Context;
import org.extism.sdk.manifest.Manifest;
import org.extism.sdk.wasm.WasmSourceResolver;

import java.nio.file.Path;

public class MainVerticle extends AbstractVerticle {

  @Override
  public void start(Promise<Void> startPromise) throws Exception {

    var resolver = new WasmSourceResolver();
    var manifest = new Manifest(resolver.resolve(Path.of("./hello-world/hello-world.wasm")));

    vertx.createHttpServer().requestHandler(req -> {

      try (var ctx = new Context(); var plugin = ctx.newPlugin(manifest, true))
      {
        var output = plugin.call("helloWorld", "Bob Morane");

        req.response()
        .putHeader("content-type", "text/plain; charset=utf-8")
        .end(output);

      }

    }).listen(8888, http -> {
      if (http.succeeded()) {
        startPromise.complete();
        System.out.println("HTTP server started on port 8888");
      } else {
        startPromise.fail(http.cause());
      }
    });


  }
}

File: samples/extism-vert-x-kotlin/src/main/kotlin/garden/bots/starter/MainVerticle.kt
package garden.bots.starter

import io.vertx.core.AbstractVerticle
import io.vertx.core.Promise

import org.extism.sdk.Context
import org.extism.sdk.manifest.Manifest
import org.extism.sdk.wasm.WasmSourceResolver
import java.nio.file.Path
import kotlin.system.exitProcess


class MainVerticle : AbstractVerticle() {

  
  override fun start(startPromise: Promise<Void>) {

    val getPlugin = {
      try {
        val resolver = WasmSourceResolver()
        val manifest = Manifest(resolver.resolve(Path.of("./hello-world/hello-world.wasm")))
        val ctx = Context() 
        ctx.newPlugin(manifest, true)
      } catch(e: Exception) {
        exitProcess(0)
      }
    }

    val plugin = getPlugin()

    vertx
      .createHttpServer()
      .requestHandler { req ->

        try {
          val output = plugin.call("helloWorld", "Bob Morane");
          req.response()
          .putHeader("content-type", "text/plain; charset=utf-8")
          .end(output)

        } catch(exception: Exception) {
          req.response()
          .putHeader("content-type", "text/plain; charset=utf-8")
          .end("Ouch 😡")
        }

      }
      .listen(8888) { http ->
        if (http.succeeded()) {
          startPromise.complete()
          println("HTTP server started on port 8888")
        } else {
          startPromise.fail(http.cause());
        }
      }
  }
}
